### ISSUE TITLE
update drops terms from ziformula

### ISSUE THREAD

**Issue by mebrooks** [glmmTMB/glmmTMB#341]
It seems that using `update` with `ziformula` causes any random effects in `ziformula` to be dropped. This happens in `m2` below, but not in `m3`.

```
m1=glmmTMB(count~mined, zi=~mined+(1|site), family="poisson", Salamanders)
m2=update(m1, ziformula=~-mined)
summary(m2) #no RE

m3=update(m1, formula=~-mined)
summary(m3) #with RE
```
I found this because of #235

**Comment by mebrooks** [comment]
I just realized that it happens for the conditional model also.
```
m4=glmmTMB(count~mined+(1|site), zi=~mined, family="poisson", Salamanders)
update(m4, formula=~-mined)
```

**Comment by koenvanbenthem** [comment]
In fact, the problem is not specific to random effects. Fixed effects are dropped too:
```
m1=glmmTMB(count~mined, zi=~mined+site, family="poisson", Salamanders)
m2=update(m1, ziformula=~-mined)
summary(m2) #no RE
```
To be more surprised: right now update just replaces the whole ziformula with whatever argument is given to it. 

Right now, at least on my machine, update.default is used. The argument is not recognized as 'the formula', so instead of updating the formula (update.formula), it runs:
```
call[[a]] <- extras[[a]]
```
with in these case a = 'ziformula'.

Maybe we should write a seperate update function for glmmTMB?
- I could give this a try.

**Comment by mebrooks** [comment]
I see that behavior too. 

I looked at some examples and realized that I was using `update` incorrectly (there were no warnings).  Using it (in a way that should be) correctly, I think it still drops terms from `ziformula` but keeps them in `formula`.

```
m1=glmmTMB(count~mined, zi=~mined+site, family="poisson", Salamanders)
m2=update(m1, ziformula=~.-mined)
summary(m2) #zi model is intercept only
```

I think it works correctly to update the conditional model. 
```
m4=glmmTMB(count~mined+(1|site), zi=~mined, family="poisson", Salamanders)
m5=update(m4, .~.-mined)
summary(m5)
```

**Comment by bbolker** [comment]
@mebrooks , in your example I think you meant `formula = ~ . - mined` ?  You need `.` in there to represent "previous version of the formula".  Otherwise `~ -mined` means "intercept-only formula, minus `mined`", which is weird but does logically give `1`

**Comment by koenvanbenthem** [comment]
something like this (?)


update.glmmTMB <- function (object, formula., ziformula., dispformula.,..., evaluate = TRUE) 
{
  if (is.null(call <- getCall(object))) 
    stop("need an object with call component")
  extras <- match.call(expand.dots = FALSE)$...
  if (!missing(formula.)) 
    call$formula <- update.formula(formula(object), formula.)
  
  if(!missing(ziformula.))
    call$ziformula <- update.formula(call$ziformula, ziformula.)
  
  
  if(!missing(dispformula.))
    call$dispformula <- update.formula(call$dispformula, dispformula.)
  
  if (length(extras)) {
    existing <- !is.na(match(names(extras), names(call)))
    for (a in names(extras)[existing]) call[[a]] <- extras[[a]]
    if (any(!existing)) {
      call <- c(as.list(call), extras[!existing])
      call <- as.call(call)
    }
  }
  if (evaluate) 
    eval(call, parent.frame())
  else call
}

**Comment by mebrooks** [comment]
I agree that I was using it wrong, but it is confusing. It's always taking `.` as the old `formula` but not the old `ziformula`.  Now, I realize this is documented in the `glmmTMB` help file so it's not exactly a bug. "Specifying ~. will set the right-hand side of the zero-inflation formula identical to the right-hand side of the main (conditional effects) formula; terms can also be added or subtracted."

```
m6=glmmTMB(count~mined, zi=~mined+site+(1|spp), family="poisson", Salamanders)
update(m6, ziformula=~.-site) #same as zi=~mined
update(m6, ziformula=~.-mined) #same as zi=~1
```
I guess the bottom line is that if you want to update `ziformula`, you have to start from scratch rather than adding or removing terms.

**Comment by mebrooks** [comment]
...or use @koenvanbenthem 's function to update `ziformula` independently of `formula`.

**Comment by bbolker** [comment]
I think we should consider changing this - Koen's function makes more sense than the current behaviour -- but it's another API-breaking change, so we have to spend some more thinking about it (how will we warn users of the change?)

**Comment by koenvanbenthem** [comment]
Maybe something like this might have less influence on the API? (I haven't really tested this version though, so not sure if it's properly functioning)
```
update.glmmTMB <- function (object, formula.,..., evaluate = TRUE) 
{
  if (is.null(call <- getCall(object))) 
    stop("need an object with call component")
  
  extras <- match.call(expand.dots = FALSE)$...
 
  if (!missing(formula.)) 
    call$formula <- update.formula(formula(object), formula.)
  
  if('ziformula' %in% names(extras))
    call$ziformula <- update.formula(call$ziformula, extras$ziformula)
  
  
  if('dispformula' %in% names(extras))
    call$dispformula <- update.formula(call$dispformula, extras$dispformula)
  
  extras <- extras[!names(extras) %in% c('dispformula','ziformula')]
  
  if (length(extras)) {
    existing <- !is.na(match(names(extras), names(call)))
    for (a in names(extras)[existing]) call[[a]] <- extras[[a]]
    if (any(!existing)) {
      call <- c(as.list(call), extras[!existing])
      call <- as.call(call)
    }
  }
  if (evaluate) 
    eval(call, parent.frame())
  else call
}
```

**Comment by mebrooks** [comment]
Both versions of `update.glmmTMB` work as I would expect them to on my computer.

If we change how `update` behaves, would we also change how `glmmTMB(..., zi=~.)` behaves? I'm OK with changing both if it make sense to do so.

**Comment by bbolker** [comment]
I'm not sure.

**Comment by mmaechler** [comment]
- I agree (with some) that we should or even *must* introduce `update.glmmTMB()`.
- If we have documented that  `zi=~.`  means to have `zi` use the same terms as the main `formula`, I'd tend to keep that, *or* introduce a new syntax to keep that (??)
- *and* carefully think on how  update.glmmTMB() should deal with different formulas, notably what "." should mean....
but I acknowledge I have not looked at the current implementation yet.

**Comment by neurotakuko** [comment]
Thank you @koenvanbenthem for this solution to unexpected behavior for ziformula updates. It took me some time to notice this since I didn't think to check for a difference in behavior for using "." to update the main and zi formulas.
